plugins/catalog-backend/src/service/AuthorizedEntitiesCatalog.ts

```diff
@@ -70,8 +72,84 @@ export class AuthorizedEntitiesCatalog implements EntitiesCatalog {
     return this.entitiesCatalog.removeEntityByUid(uid);
   }
 
-  entityAncestry(entityRef: string): Promise<EntityAncestryResponse> {
-    // TODO: Implement permissioning
-    return this.entitiesCatalog.entityAncestry(entityRef);
+  async entityAncestry(
+    entityRef: string,
+    options?: { authorizationToken?: string },
+  ): Promise<EntityAncestryResponse> {
+    const rootEntityAuthorizeResponse = (
+      await this.permissionApi.authorize(
+        [{ permission: catalogEntityReadPermission, resourceRef: entityRef }],
+        { token: options?.authorizationToken },
+      )
+    )[0];
+    if (rootEntityAuthorizeResponse.result === AuthorizeResult.DENY) {
+      throw new NotAllowedError();
+    }
+
+    const ancestryResult = await this.entitiesCatalog.entityAncestry(entityRef);
+    const authorizeResponse = await this.permissionApi.authorize(
+      ancestryResult.items.map(item => ({
+        permission: catalogEntityReadPermission,
+        resourceRef: stringifyEntityRef(item.entity),
+      })),
+      { token: options?.authorizationToken },
+    );
+    const unauthorizedAncestryItems = ancestryResult.items.filter(
+      (_, index) => authorizeResponse[index].result === AuthorizeResult.DENY,
+    );
+    const rootUnauthorizedEntityRefs = unauthorizedAncestryItems.map(
+      ancestryItem => stringifyEntityRef(ancestryItem.entity),
+    );
+    const allUnauthorizedEntityRefs = new Set(
+      rootUnauthorizedEntityRefs
+        .map(rootEntityRef =>
+          this.findUnauthorizedParents(
+            rootEntityRef,
+            ancestryResult.items,
+            new Set(rootUnauthorizedEntityRefs),
+          ),
+        )
+        .flat(),
+    );
+    return {
+      rootEntityRef: ancestryResult.rootEntityRef,
+      items: ancestryResult.items.filter(
+        ancestryItem =>
+          !allUnauthorizedEntityRefs.has(
+            stringifyEntityRef(ancestryItem.entity),
+          ),
+      ),
+    };
+  }
+
+  private findUnauthorizedParents(
```

Comment:
Are there actually assumptions in this method about the parents being unauthorized? I think this is a nice method to break out, but just wonder if it might be better named `findParents`.